Fix service address binding

gaussianrecurrence opened the pull request:
- Initially the uSVC was bound to 127.0.0.1 which worked for local tests. But:
   * Not having it listen to all interfaces would make it fail whenever setting it up as a pod/container.
   * Also, it wouldn't allow IPv6 binding.
 - Consequently bind address has been changed to [::]:8000 and if it fails due to lack of support for IPv6, it falls back to
   0.0.0.0:8000
 - Added support for `EAFNOSUPPORT` which is returned by `socket` [1] whenever
   IPv6 is not supported in the environment.
 - The log indicating the uSVC is falling back to IPv4 now uses `log::warn` [2] for
   consistency with actix logs.
 - Added necessary `log` dependency v0.4.27 [3].

Issue: #7

[1]: https://man7.org/linux/man-pages/man2/socket.2.html
[2]: https://docs.rs/log/0.4.27/log/macro.warn.html
[3]: https://docs.rs/log/0.4.27/log/

gaussianrecurrence committed:
Fix service address binding

 - Initially the uSVC was bound to 127.0.0.1 which worked for local tests. But:
   * Not having it listen to all interfaces would make it fail whenever
     setting it up as a pod/container.
   * Also, it wouldn't allow IPv6 binding.
 - Consequently bind address has been changed to [::]:8000 and if it fails due
   to lack of support for IPv6, it falls back to 0.0.0.0:8000

Issue: #7

msdt93 commented:
In order to verify IPv4 fallback, I tricked the uSVC to think IPv6 is not available with the following hooks:

#### block_ipv6_socket.c
```c
// compile with gcc -shared -fPIC -o block_ipv6_socket.so block_ipv6_socket.c -ldl

#define _GNU_SOURCE
#include <dlfcn.h>
#include <sys/socket.h>
#include <errno.h>

typedef int (*socket_func_t)(int domain, int type, int protocol);
static socket_func_t real_socket = 0;

int socket(int domain, int type, int protocol) {
    if (!real_socket) {
        real_socket = (socket_func_t)dlsym(RTLD_NEXT, "socket");
    }

    if (domain == AF_INET6) {
        errno = EAFNOSUPPORT; // Address family not supported
        return -1;
    }

    return real_socket(domain, type, protocol);
}
```

#### block_ipv6_bind.c
```c
// compile with gcc -shared -fPIC -o block_ipv6_bind.so block_ipv6_bind.c -ldl

#define _GNU_SOURCE
#include <dlfcn.h>
#include <sys/socket.h>
#include <errno.h>

typedef int (*bind_func_t)(int sockfd, const struct sockaddr *addr, socklen_t addrlen);
static bind_func_t real_bind = 0;

int bind(int sockfd, const struct sockaddr *addr, socklen_t addrlen) {
    if (!real_bind) {
        real_bind = (bind_func_t)dlsym(RTLD_NEXT, "bind");
    }

    if (addr && addr->sa_family == AF_INET6) {
        errno = EADDRNOTAVAIL; // Address not available
        return -1;
    }

    return real_bind(sockfd, addr, addrlen);
}
```
---
For reference, the hooks can be used as follow:
```shell
$ LD_PRELOAD="<path_to_hook_1.so> ··· <path_to_hook_N.so>" <application>
```

**E.g:**
```shell
$ LD_PRELOAD="block_ipv6_socket.so block_ipv6_bind.so" target/release/chattopia-api
```

msdt93 commented:
Trying to verify IPv4 fallback in a1fa978, an issue was detected. Whenever hooking socket call to return `EAFNOSUPPORT` for IPv6 this can be observed:
```shell
$ LD_PRELOAD="$(realpath block_ipv6_socket.so)" target/release/chattopia-api
Error: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" }
```

However, if only bind is hooked, the IPv4 fallback works, as can be seen:
```
$ LD_PRELOAD="$(realpath block_ipv6_bind.so)" target/release/chattopia-api
[2025-05-20T08:50:59Z WARN  chattopia_api] IPv6 not available, falling back to IPv4
[2025-05-20T08:50:59Z INFO  actix_server::builder] starting 8 workers
[2025-05-20T08:50:59Z INFO  actix_server::server] Actix runtime found; starting in Actix runtime
[2025-05-20T08:50:59Z INFO  actix_server::server] starting service: "actix-web-service-0.0.0.0:8000", workers: 8, listening on: 0.0.0.0:8000
```

gaussianrecurrence committed:
Add support for EAFNOSUPPORT error

 - Added support for EAFNOSUPPORT which is returned by `socket` [1] whenever
   IPv6 is not supported in the environment.

[1]: https://man7.org/linux/man-pages/man2/socket.2.html

gaussianrecurrence committed:
Change fallback log to use log::warn macro instead

 - The log indicating the uSVC is falling back to IPv4 now uses `log::warn` [1] for
   consistency with actix logs.
 - Added necessary `log` dependency v0.4.27 [2].

[1]: https://docs.rs/log/0.4.27/log/macro.warn.html
[2]: https://docs.rs/log/0.4.27/log/

msdt93 commented:
Verifying IPv4 fallback in 710b1ba, it seems to be working fine.

#### `socket` hooked
```shell
$ LD_PRELOAD="$(realpath block_ipv6_socket.so)" target/release/chattopia-api
[2025-05-20T09:09:31Z WARN  chattopia_api] IPv6 not available, falling back to IPv4
[2025-05-20T09:09:31Z INFO  actix_server::builder] starting 8 workers
[2025-05-20T09:09:31Z INFO  actix_server::server] Actix runtime found; starting in Actix runtime
[2025-05-20T09:09:31Z INFO  actix_server::server] starting service: "actix-web-service-0.0.0.0:8000", workers: 8, listening on: 0.0.0.0:8000
```

#### `bind` hooked
```shell
$ LD_PRELOAD="$(realpath block_ipv6_bind.so)" target/release/chattopia-api
[2025-05-20T09:10:16Z WARN  chattopia_api] IPv6 not available, falling back to IPv4
[2025-05-20T09:10:16Z INFO  actix_server::builder] starting 8 workers
[2025-05-20T09:10:16Z INFO  actix_server::server] Actix runtime found; starting in Actix runtime
[2025-05-20T09:10:16Z INFO  actix_server::server] starting service: "actix-web-service-0.0.0.0:8000", workers: 8, listening on: 0.0.0.0:8000
```

#### `socket` & `bind` hooked
```shell
$ LD_PRELOAD="$(realpath block_ipv6_bind.so) $(realpath block_ipv6_socket.so)" target/release/chattopia-api
[2025-05-20T09:11:03Z WARN  chattopia_api] IPv6 not available, falling back to IPv4
[2025-05-20T09:11:03Z INFO  actix_server::builder] starting 8 workers
[2025-05-20T09:11:03Z INFO  actix_server::server] Actix runtime found; starting in Actix runtime
[2025-05-20T09:11:03Z INFO  actix_server::server] starting service: "actix-web-service-0.0.0.0:8000", workers: 8, listening on: 0.0.0.0:8000
```

#### No hooks with IPv6 support
```shell
$ target/release/chattopia-api
[2025-05-20T09:11:38Z INFO  actix_server::builder] starting 8 workers
[2025-05-20T09:11:38Z INFO  actix_server::server] Actix runtime found; starting in Actix runtime
[2025-05-20T09:11:38Z INFO  actix_server::server] starting service: "actix-web-service-[::]:8000", workers: 8, listening on: [::]:8000
```
